### ISSUE TITLE
Issue handling full disk

### ISSUE THREAD

**Issue by cbalster** [osmdroid/osmdroid#1171]
## Issue Type

[X] Bug

## Description and/or steps/code to reproduce the problem

When the volume the tilecache is saved on is full and the map is scrolled to a location with tiles not in the cache there are recurrent `SQLiteFullException`s and infinite redownloads of the corresponding maptiles:
![capture](https://user-images.githubusercontent.com/9266843/47255690-5a892680-d475-11e8-8524-d1079b14210c.PNG)

`E/SQLiteDatabase: Error inserting expires=1540145732000 key=204064317 tile=[B@29635de provider=Mapnik
    android.database.sqlite.SQLiteFullException: database or disk is full (code 13)
        at android.database.sqlite.SQLiteConnection.nativeExecuteForLastInsertedRowId(Native Method)
        at android.database.sqlite.SQLiteConnection.executeForLastInsertedRowId(SQLiteConnection.java:782)
        at android.database.sqlite.SQLiteSession.executeForLastInsertedRowId(SQLiteSession.java:788)
        at android.database.sqlite.SQLiteStatement.executeInsert(SQLiteStatement.java:86)
        at android.database.sqlite.SQLiteDatabase.insertWithOnConflict(SQLiteDatabase.java:1471)
        at android.database.sqlite.SQLiteDatabase.replace(SQLiteDatabase.java:1387)
        at org.osmdroid.tileprovider.modules.SqlTileWriter.saveFile(SqlTileWriter.java:149)
        at org.osmdroid.tileprovider.modules.TileDownloader.downloadTile(TileDownloader.java:152)
        at org.osmdroid.tileprovider.modules.MapTileDownloader$TileLoader.downloadTile(MapTileDownloader.java:210)
        at org.osmdroid.tileprovider.modules.MapTileDownloader$TileLoader.loadTile(MapTileDownloader.java:239)
        at org.osmdroid.tileprovider.modules.MapTileModuleProviderBase$TileLoader.run(MapTileModuleProviderBase.java:312)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
        at java.lang.Thread.run(Thread.java:818)`

### If it's a bug, version(s) of android this affects:

Observed on API21 - probably all others too

### Version of osmdroid the issue relates to:

6.0.2

**Comment by monsieurtanuki** [comment]
@cbalster The garbage collector in `SqlTileWriter` is supposed to work asynchronously, which would explain a non immediate action - but not doing nothing during several minutes.

Perhaps the key would be to refresh the database: could you try to make `SqlTileWriter.isFunctionalException` return `false` for  `"SQLiteFullException"`? (which would make sense anyway)

**Comment by cbalster** [comment]
Finally found some time to look at this. Besides the fact that there is no trimming happening even though space is below the trim limit, I noticed the following:
This https://github.com/osmdroid/osmdroid/blob/5e710580c2c948fd1c0e76f0c2e52b291c472d6e/osmdroid-android/src/main/java/org/osmdroid/tileprovider/modules/SqlTileWriter.java#L156 catch block is never reached, since `db.replace(TABLE, null, cv);` at https://github.com/osmdroid/osmdroid/blob/5e710580c2c948fd1c0e76f0c2e52b291c472d6e/osmdroid-android/src/main/java/org/osmdroid/tileprovider/modules/SqlTileWriter.java#L149 doesn't throw but just returns -1 in case of an exception:
```
    public long replace(String table, String nullColumnHack, ContentValues initialValues) {
        try {
            return insertWithOnConflict(table, nullColumnHack, initialValues,
                    CONFLICT_REPLACE);
        } catch (SQLException e) {
            Log.e(TAG, "Error inserting " + initialValues, e);
            return -1;
        }
    }
```

**Comment by monsieurtanuki** [comment]
@cbalster Good catch! The idea would be to use `replaceOrThrow` instead of `replace`, right? Or `insertWithOnConflict`? (I coudn't find the detailed source code for `SQLiteDatabase`)

**Comment by cbalster** [comment]
Okay, so I tested this and it obviously throws as expected now with `replaceOrThrow()`. The Problem is that once there is so little space that we get a `SQLiteFullException` there, we basically won't be able to do anything with the db at all since even the `db.delete()` in `runCleanupOperation()` will also throw the same exception. Basically this means this is an error that cannot be recovered from without the user manually freeing up space. The only option from inside osmdroid I see, would be to delete the whole database which I think is not really an option.
Imho this means there has to be a mechanism which prevents completely filling up the disk. Without looking into it to much I see two options here:
1) Add a check in `SqlTileWriter` that always sets a hard-limit if getTileFileSystemCacheMaxBytes() returns anything that would bring the remaining free space under e.g. 5MB.
2) Use the `schema.max_page_count = N`  pragma to limit the maximum db size to something equivalent as in 1). Since this would require handling the max page count reached case, I'd prefer 1)

Another problem is the time based triggering of the gc in `saveTile()`. This opens a time window in which  the disk can be filled without a GC occuring. An option here would be to move the size check from `runCleanupOperation()` to the time check in `saveTile()`. This might degrade performance though and maybe would require some testing.

As a general Note: All this only really becomes apparent when the DB + remaining space is quite small (< ~25 MB) and especially if the default max cache and trim sizes (0.95 and 0.90) are used, as this results in rather little free space before the GC kicks in.

**Comment by cbalster** [comment]
Actually option number 2) might be the better one: Restricting page count should also result in `SQLiteFullException` s if the size is exceeded. _If_ the space required by `delete()` doesn't count against this limit this might be a nice solution.

**Comment by monsieurtanuki** [comment]
Not even being able to delete when the database is full? That's not good news!
(actually the term "database is full" seems to mean more something like "filesystem where the database file is is full")

Well, I don't think that we can do much about the database being full without degrading substantially the performances.
And it may even be useless: we can put many checks _about the data inside the database itself_, but what happens if at the same time the user/another app downloads a file on the same filesystem? We'll get the same SQLite Full Exception, regardless of how careful we were.

Your issue was initially about the endless (and eventually pointless) downloads that happen after a SQLite Full Exception.
As I assume that we cannot be smart enough to deal with osmdroid tile data sizes and the filesystem at the same time, I think the fix would be somewhere else.

Recently, I worked on #989, where there were endless download retries when the tile server was temporarily unreachable. The solution was to say "oh if it's unreachable then I'll try a little later", with progressive delays.

For this issue, the idea could be to say that if ever we catch a SQLite Full Exception, we won't try to download anything before a delay. After this delay, we can run the garbage collector, enable the downloads again, and wait a little longer if we still have the SQLite Full Exception. We could even warn the end user with a `Toast`, something like a `SQLiteFullExceptionListener` callback.

@cbalster What do you think of that?

**Comment by monsieurtanuki** [comment]
Merged #1197.
